dev-db/mongodb: add 6.0.21, 7.0.18, 8.0.6 and dev-python/retry (needed for mongodb-8.0.6)

fvalenduc opened the pull request:
Bug: https://bugs.gentoo.org/882515

This pull request add release 7.0.18 of mongodb, since the lastest version available in Gentoo (5.0.30) is now end-of-life (see https://www.mongodb.com/legal/support-policy/lifecycles).
Some patches are taken from the archlinux user repository (extrapatch-sconstruct.patch, mongodb-7.0.1-sconstruct.patch and dev-db/mongodb/files/mongodb-7.0.12-boost-1.85.patch). I kept some patches from the current version in gentoo (       mongodb-4.4.29-no-enterprise.patch, mongodb-5.0.2-no-compass.patch and mongodb-5.0.2-skip-reqs-check.patch). They still applies, but I am not sure these are still necessary. Finally, I wrote the 2 last one myself taking inspiration from other patches dealing with similar issues in newer versions of boost (0001-add-boost-include.patch and 0001-replace-change_extension-by-replace_extension.patch.

I also tested this with python 3.13 and it compiles and work correctly, so I set this in the ebuild:
PYTHON_COMPAT=( python3_{9..13} )
Should I restrict the versions ?

Thanks for your attention.

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

Dessa commented:
unfortunately, we would still have to provide 6.0 for upgraders, since we can't assume everybody is starting fresh.

``To upgrade from a version earlier than the 6.0-series, you must successively upgrade major releases until you have upgraded to 6.0-series. For example, if you are running a 5.0-series, you must upgrade first to 6.0 before you can upgrade to 7.0.``

is this based of off an older ebuild? there were some changes to it since which seem to be lost here

thank you for the boost fixes, these were my personal blocker getting this going (well, and upstream not releasing sources for 2 months...)
i don't think we can unbundle libbson or mongo-c-driver yet due to not matching versions; i'd be fine with having a higher version, which is what will end up happening when the security bug is dealt with, but not a lower.

these 4 newly added python deps seem suspicious, does the package really not build / work without them?

i'll give this a go with my usual build testing tomorrow either way.

fvalenduc commented:
Thanks for the comment. I don't think memory-profiler and cxxflt are necessary, and these are not available in gentoo. I am trying to built without these to confirm this.

fvalenduc commented:
I removed memory-profiler and cxxflt and it still compiles. I also restarted from the current ebuild for 5.0.30. I removed all patches which do not apply and I also add to remove --system-pcre in the arguments of scons otherwise it does not work.

fvalenduc commented:
So, I finally managed to solve the problems raised by the CI reports. I tried to add mongodb-6.0.21, unfortunately without success. It fails with this error:

src/mongo/util/processinfo_linux.cpp: In function 'void mongo::{anonymous}::appendMountInfo(mongo::BSONObjBuilder&)':
src/mongo/util/processinfo_linux.cpp:302:22: error: variable 'std::ifstream ifs' has initializer but incomplete type
  302 |     std::ifstream ifs("/proc/self/mountinfo");

I have no idea on how to solve this.

fvalenduc commented:
I managed to get a working ebuild for 6.0.21, so I added it to the pull request.

fvalenduc commented:
So the checks failed because dev-python/retry is not available. However I have included it in the pull request. Is this expected ? Or is there a way to solve this ?
According to the PKGBUILD file in archlinux user repository, retry is well needed.

thesamesam commented:
It currently has CI failing and you already had a reply from @Dessa. I can take a look but please avoid remarks like that.

thesamesam reviewed (commented):
I guess I can look at taking the Python package for python@ if its testsuite works. Could you maybe post an inline diff as a comment (with backticks around it) for the mongodb ebuilds to make them easier to review?

Dessa commented:
Currently not near my computer and gonna be limited for a bit too, but I noticed another thing with retry - it fails when pbr is installed. "reretry" is in science overlay which could be an alternative, but I had only tested itself and not if it actually works with mongodb yet and I'm pretty much on a time constraint until next Saturday

fvalenduc commented on dev-python/retry/retry-0.9.2.ebuild:
```diff
@@ -0,0 +1,33 @@
+SLOT="0"
+KEYWORDS="amd64 ~arm ~arm64 ~ppc ~ppc64 ~riscv ~s390 ~sparc x86"
+
+python_test() {
```
It is my first pull request ever... And I do not have any experience at all at writing python ebuild. So I copied another python ebuild and I adapted it for retry. So feel free to adapt the ebuild. It compiles and install, but there might be problems. I only added it because it seems to be needed for mongodb 8.

thesamesam replied:
I'll try find some time to handle the Python bits over the weekend.

fvalenduc commented:
The only reason for which the CI fails is that it needs dev-python/retry which is not in portage currently.

thesamesam commented:
(Note that the reason CI is failing is the keywords, not the fact the package is new -- you can have clean CI for a PR which introduces new packages.)

fvalenduc commented:
Thanks for the hint. I adapted the keywords like this for retry:
KEYWORDS="~amd64 ~arm ~arm64 ~ppc ~ppc64 ~riscv ~s390 ~sparc ~x86"

So I am curious for the result now.

fvalenduc commented:
Sorry I messed up with git and the manifest file is broken...
I will solve it later today hopefully.

fvalenduc commented:
The last version of the pull request should fix the issue found in mongodb. I have also verified and retry is really needed. Without it, build immediately fails:

ModuleNotFoundError: No module named 'retry':
  File "/var/tmp/notmpfs/portage/dev-db/mongodb-8.0.6/work/mongo-r8.0.6/SConstruct", line 6690:
    env.Tool('integrate_bazel')
  File "/var/tmp/notmpfs/portage/dev-db/mongodb-8.0.6/work/mongo-r8.0.6/src/third_party/scons-3.1.2/scons-local-3.1.2/SCons/Environment.py", line 1822:
    tool = SCons.Tool.Tool(tool, toolpath, **kw)
  File "/var/tmp/notmpfs/portage/dev-db/mongodb-8.0.6/work/mongo-r8.0.6/src/third_party/scons-3.1.2/scons-local-3.1.2/SCons/Tool/__init__.py", line 122:
    module = self._tool_module()
  File "/var/tmp/notmpfs/portage/dev-db/mongodb-8.0.6/work/mongo-r8.0.6/src/third_party/scons-3.1.2/scons-local-3.1.2/SCons/Tool/__init__.py", line 239:
    module = spec.loader.load_module(spec.name)
  File "<frozen importlib._bootstrap_external>", line 680:
    
  File "<frozen importlib._bootstrap_external>", line 1207:
    
  File "<frozen importlib._bootstrap_external>", line 1031:
    
  File "<frozen importlib._bootstrap>", line 537:
    
  File "<frozen importlib._bootstrap>", line 966:
    
  File "<frozen importlib._bootstrap>", line 935:
    
  File "<frozen importlib._bootstrap_external>", line 1026:
    
  File "<frozen importlib._bootstrap>", line 488:
    
  File "/var/tmp/notmpfs/portage/dev-db/mongodb-8.0.6/work/mongo-r8.0.6/site_scons/site_tools/integrate_bazel.py", line 19:
    from retry import retry

fvalenduc commented:
So I finally manage to solve all errors reported by the QA checks. If somebody want to improve the ebuild for retry, please do.
I hope it can now be taken into consideration.

fvalenduc commented:
Following the suggestion of @dessa, I replaced retry by reretry and it works too. It's probably a much better choice because there has been no updates to the github repository since 9 years. 
The ebuild in the science overlay has support up to python 3.12 but 3.13 works also. 
I don't know how to proceed. Can I takeover the reretry ebuild and include it in this pull request ?

fvalenduc commented:
In fact, retry is only used once in integrate_bazel.py. We can just as well remove it.

fvalenduc commented:
I upgraded mongo 6 to 6.0.22 and mongo 8 to 8.0.8. I also propose using tenacity which is already in gentoo instead of retry or reretry for 8.0.8 (or skipping it).
I hope this will one day be taken in consideration.

thesamesam commented:
Sure, just had no time to look yet. I'm not sure what you mean wrt tenacity/retry/reretry. Can you explain the background there? Did upstream migrate which ones they need?

fvalenduc commented:
integrate_bazel.py contains this (in /site_scons/site_tools):

@retry(tries=5, delay=3)
def download_path_with_retry(*args, **kwargs):
    urllib.request.urlretrieve(*args, **kwargs)

So either you remove retry in the decorator or you replace it by this to use tenacity:
retry(stop=stop_after_attempt(3), wait=wait_fixed(0.1))
And of course, you modify the import line in the beginning of the file. That is what the patch "0001-use-tenacity" is doing.

fvalenduc commented:
I also tried to compile it with clang, unfortunately without success.  I adapted some patches of the 5.0.30 version, but it is not enough.
I archlinux, it does not work either. Is it a requirement, or a nice to have ?

thesamesam commented on dev-db/mongodb/files/mongodb-8.0.6-skip-reqs-check.patch:
```diff
@@ -0,0 +1,24 @@
+diff --git a/buildscripts/scons2.py b/buildscripts/scons.py
```
Commentary.

thesamesam reviewed (commented):
Left some brief comments. It looks pretty good at a glance. I haven't diffed the build system at all yet. I'll do that hopefully tomorrow but over the weekend if not.

Thanks for your persistence on this. I do appreciate it.

fvalenduc commented on dev-db/mongodb/files/0001-use-tenacity.patch:
```diff
@@ -0,0 +1,34 @@
+From 1d2201e3657c8487bef828b904dae87868b142bf Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?Fran=C3=A7ois=20Valenduc?= <francoisvalenduc@gmail.com>
+Date: Tue, 29 Apr 2025 11:39:36 +0200
+Subject: [PATCH] use tenacity
```
Sure. It depends what we prefer. Either we keep with the original code. Then we need to add the retry python package (see above). This patch keep the retry decorator but replace it by tenacity which is already available in gentoo.
I can also add a patch which remove retry.

thesamesam replied:
Thanks. So, it's just about making it easier to get in (which is reasonable) given it's a new dep for something we can easily replace with a tiny patch. Got it.

fvalenduc commented:
So here an updated version. I updated the python targets (3.11 to 3.13) and added a comment to the tenacity patch.

thesamesam commented:
Thanks!
